jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,395 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A fast container of poolable objects, with optional support for
+ * multiplexing and thread-local caching.
+ * @param <T>
+ */
+public class Pool<T> implements AutoCloseable
+{
+    private static final Logger LOGGER = Log.getLogger(Pool.class);
+
+    private final List<Entry> sharedList = new CopyOnWriteArrayList<>();
+    private final Collection<List<Entry>> allCaches = new CopyOnWriteArrayList<>();
+    private final ThreadLocal<List<Entry>> cache;
+    private final Lock closeLock = new ReentrantLock();
```

Comment:
Not sure what this lock protects. Apparently not field `closed` as it's `volatile` and often read outside the lock?

Reply:
The lock is only about preventing concurrent `add()` and `close()` calls, i.e.: don't allow adding new objects to the pool once `close()` has been called. I've reworked the closed state to check it from all methods and added some extra test.